### PR TITLE
Implement a query based producer

### PR DESCRIPTION
**Pull request by fernandofcampos**

### CODE REVIEW

File: app/domain/service.go [hunk 1]
```diff
@@ -18,3 +19,15 @@ type ProcessorService interface {
 type FilterInterface[T any] interface {
 	ShouldProcess(typeValue *T) bool
 }
+
+type QueryProcessor interface {
+	ProcessQueries(ctx context.Context) error
+}
+
+type QueryService interface {
+	GetLatestNetworkInferences(ctx context.Context, topicID uint64) (*json.RawMessage, error)
```
Comment (spooktheducks) [inline app/domain/service.go:28]: `json.RawMessage` is already a `[]byte` so it's not really necessary to have a pointer to it

File: app/usecase/queryproducer.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		// Process queries for the current epoch
+		if err := q.processor.ProcessQueries(ctx); err != nil {
+			q.logger.Warn().Err(err).Msg("Failed to process queries")
+			// Sleep to avoid spamming the node
+			time.Sleep(q.rateLimitInterval)
+			continue
+		}
+
+		// Sleep until the next epoch
+		time.Sleep(q.epochInterval)
```
Comment (spooktheducks) [inline app/usecase/queryproducer.go:48]: Does this backfill if it misses an epoch for any reason?

File: app/usecase/queryproducer.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	if processor == nil {
+		return nil, fmt.Errorf("processor is nil")
+	}
+
+	logger := log.With().Str("producer", "query").Logger()
+
+	return &QueryProducer{processor: processor, epochInterval: epochInterval, rateLimitInterval: rateLimitInterval, logger: &logger}, nil
+}
+
+func (q *QueryProducer) Execute(ctx context.Context) error {
```
Comment (spooktheducks) [inline app/usecase/queryproducer.go:30]: For simplicity's sake, I would probably just merge this function into `QueryProcessor`

File: cmd/producer/main.go [hunk 4]
```diff
@@ -136,13 +135,42 @@ func initializeApp(ctx context.Context, cfg config.Config) (*app.App, func(), er
...
+	if err != nil {
+		return nil, nil, err
+	}
+
+	queryProducer, err := usecase.NewQueryProducer(queryProcessor, cfg.QueryProducer.EpochInterval, cfg.QueryProducer.RateLimitInterval)
+	if err != nil {
+		return nil, nil, err
+	}
+
 	cleanup := func() {
```
Comment (spooktheducks) [inline cmd/producer/main.go:163]: If one of the previous statements returns an error, won't we be leaking DB connections?

**Review by spooktheducks — CHANGES_REQUESTED** [review]

File: cmd/producer/main.go (hunk 4 above)
Comment (fernandofcampos) [reply cmd/producer/main.go:163]: None of those Close() methods return error.

File: app/domain/service.go (hunk 1 above)
Comment (fernandofcampos) [reply app/domain/service.go]: Done

File: app/usecase/queryproducer.go (hunk 2 above)
Comment (fernandofcampos) [reply app/usecase/queryproducer.go:48]: Not sure. In theory epoch interval is much larger than the execution time of each iteration.

File: app/usecase/queryproducer.go (hunk 3 above)
Comment (fernandofcampos) [reply app/usecase/queryproducer.go:30]: I wanted to separate coordination from processing, so we can sophisticate this in the future.

File: app/service/query.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+		TopicId: topicID,
+	}, &resp)
+}
+
+func (q *QueryService) GetNaiveInfererNetworkRegret(ctx context.Context, topicID uint64) (json.RawMessage, error) {
+	var resp emissionTypes.GetNaiveInfererNetworkRegretResponse
+	return q.QueryJSON(ctx, "/emissions.v5.QueryService/GetNaiveInfererNetworkRegret", &emissionTypes.GetNaiveInfererNetworkRegretRequest{
+		TopicId: topicID,
+	}, &resp)
+}
```
Comment (KjetilVaa) [inline app/service/query.go:73]: I thought this guy was included in the GetLatestNetworkInferences response?

File: app/service/queryprocessor.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	return nextTopicID - 1, nil
+}
+
+func (q *QueryProcessorService) ProcessNetworkInferences(ctx context.Context, topicID uint64) error {
+	networkInferences, err := q.queryService.GetLatestNetworkInferences(ctx, topicID)
+	if err != nil {
+		return fmt.Errorf("failed to get latest network inferences: %w", err)
+	}
+
+	metadata := json.RawMessage(fmt.Sprintf(`{"topic_id": %d}`, topicID))
```
Comment (fernandofcampos) [inline app/service/queryprocessor.go:48]: See if it's possible to include block height at all metadata fields.

File: app/service/query.go (hunk 5 above)
Comment (fernandofcampos) [reply app/service/query.go:73]: It is but we decided to keep both.

**Review by KjetilVaa — DISMISSED** [review]
Thanks @fernandofcampos for walking me through the this.
I think this looks good as a general framework for fetching data from API's in CRON-like manner. Some minor things we talked about:
- add block_height to the metadata field
- we need to keep in mind the versioning of the RPC endpoints. We could potentially switch to the RPC client however we would still need to update the imports for each new release.
Let's deploy this to unblock research after adding the block height

File: app/service/queryprocessor.go (hunk 6 above)
Comment (fernandofcampos) [reply app/service/queryprocessor.go]: Done

**Review by KjetilVaa — APPROVED** [review]
LGTM

File: cmd/producer/main.go (hunk 4 above)
Comment (fernandofcampos) [reply cmd/producer/main.go]: Ah, now I get it. Good catch. I've handled that case.

**Review by spooktheducks — APPROVED** [review]

**Pull request merged**